blog/_src/posts/2019-05-14-scoping-in-r.md

```diff
@@ -0,0 +1,329 @@
+    Title: Scoping in R
+    Date: 2019-05-14T10:28:00
+    Tags: scoping, r, by Ming-Ho Yee
+
+This all started with a simple question about the R programming language:
+_is R lexically or dynamically scoped?_
+
+To answer that question, we need to understand what _scoping_ means, along with
+_lexical scope_ and _dynamic scope_.
+
+<!-- more -->
+
+## What is scoping?
+
+_Scoping_ refers to the places in a program where a variable is visible and can
+be used. In particular, the interesting situation is when a function has free
+variables. Consider the example below:
+
+```r
+x <- 1
+f <- function(a) x + a
+g <- function() {
+  x <- 2
+  f(0)
+}
+g() # what does this return?
+```
+
+In this snippet, we define a variable `x` at the top level and two functions `f`
+and `g`. The body of `g` uses the parameter `a`, but also the free variable `x`.
+On line 7, we call `g`, which in turn calls `f` and returns that result.
+(In this language, the last expression of a function is implicitly returned.)
+
+What value does `g` return when it is called?
+
+To answer that question, we need to know what value `x` has on line 2. Is it
+bound to the `x` defined on line 1, in the same lexical environment that `f` was
+defined? Or is it bound to the `x` defined on line 4, in the same dynamic
+environment that `f` was called?
+
+
+### Lexical scoping
+
+Under _lexical scoping_ (also known as _static scoping_), a free variable is
+evaluated in its _lexical environment_, i.e., where it is defined. In the example
+above, the free variable `x` is used in the function `f`, which is defined at
+the top-level environment. At that point in the program, there is a definition
+of `x` visible, with value `1`.
+
+Therefore, under lexical scoping, the example program returns `1`.
+
+Most programming languages we use today are lexically scoped. In principle, a
```

Comment:
I need to rethink the wording, because saying "A compiler could statically determine what values free variables are bound to" isn't quite true; R is a counterexample.